multiplex-rs/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+    }
+}
+
+impl<T: AsyncRead> AsyncRead for Substream<T> {}
+
+impl<T: AsyncWrite> Write for Substream<T> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mut lock = match self.state.poll_lock() {
+            Async::Ready(lock) => lock,
+            Async::NotReady => return Err(io::ErrorKind::WouldBlock.into()),
+        };
+
+        let mut buffer = self.buffer
+            .take()
+            .unwrap_or_else(|| io::Cursor::new(buf_from_slice(buf)));
+
+        let out = write_stream(
+            &mut *lock,
+            write::WriteRequest::substream(MultiplexHeader::message(self.id, self.end)),
+            &mut buffer,
+        );
+
+        if buffer.position() < buffer.get_ref().len() as u64 {
+            self.buffer = Some(buffer);
+        }
+
+        out
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        let mut lock = match self.state.poll_lock() {
+            Async::Ready(lock) => lock,
+            Async::NotReady => return Err(io::ErrorKind::WouldBlock.into()),
+        };
+
+        lock.stream.flush()
+    }
+}
+
+impl<T: AsyncWrite> AsyncWrite for Substream<T> {
+    fn shutdown(&mut self) -> Poll<(), io::Error> {
+        Ok(Async::Ready(()))
+    }
+}
+
+pub struct InboundFuture<T> {
+    end: Endpoint,
+    state: Arc<Mutex<MultiplexShared<T>>>,
+}
+
+impl<T: AsyncRead> Future for InboundFuture<T> {
+    type Item = Substream<T>;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let mut lock = match self.state.poll_lock() {
+            Async::Ready(lock) => lock,
+            Async::NotReady => return Ok(Async::NotReady),
+        };
+
+        // Attempt to make progress, but don't block if we can't
+        match read_stream(&mut lock, None) {
+            Ok(_) => {}
+            Err(ref err) if err.kind() == io::ErrorKind::WouldBlock => {}
+            Err(err) => return Err(err),
+        }
+
+        let id = if let Some((id, _)) = lock.to_open.iter().next() {
+            *id
+        } else {
+            return Ok(Async::NotReady);
+        };
+
+        let name = lock.to_open.remove(&id).expect(
+            "We just checked that this key exists and we have exclusive access to the map, QED",
+        );
+
+        lock.open_stream(id);
+
+        Ok(Async::Ready(Substream::new(
+            id,
+            self.end,
+            name,
+            Arc::clone(&self.state),
+        )))
+    }
+}
+
+pub struct OutboundFuture<T> {
+    meta: Arc<MultiplexMetadata>,
+    current_id: Option<(io::Cursor<ByteBuf>, u32)>,
+    state: Arc<Mutex<MultiplexShared<T>>>,
+}
+
+impl<T> OutboundFuture<T> {
+    fn new(muxer: Multiplex<T>) -> Self {
+        OutboundFuture {
+            current_id: None,
+            meta: muxer.meta,
+            state: muxer.state,
+        }
+    }
+}
+
+fn nonce_to_id(id: usize, end: Endpoint) -> u32 {
+    id as u32 * 2 + if end == Endpoint::Dialer { 0 } else { 1 }
```

Comment:
Does that work now, or are there still conflicts with substream IDs?

Reply:
It works when we're the dialer, it _should_ work when we're the server too.